src/Umbraco.Cms.Api.Management/Factories/ContentUrlFactory.cs

```diff
@@ -0,0 +1,70 @@
+﻿using Microsoft.Extensions.Logging;
+using Umbraco.Cms.Api.Management.ViewModels.Content;
+using Umbraco.Cms.Core.Models;
+using Umbraco.Cms.Core.Models.PublishedContent;
+using Umbraco.Cms.Core.Routing;
+using Umbraco.Cms.Core.Services;
+using Umbraco.Cms.Core.Web;
+using Umbraco.Extensions;
+
+namespace Umbraco.Cms.Api.Management.Factories;
+
+public class ContentUrlFactory : IContentUrlFactory
+{
+    private readonly IPublishedRouter _publishedRouter;
+    private readonly IUmbracoContextAccessor _umbracoContextAccessor;
+    private readonly ILanguageService _languageService;
+    private readonly ILocalizedTextService _localizedTextService;
+    private readonly IContentService _contentService;
+    private readonly IVariationContextAccessor _variationContextAccessor;
+    private readonly ILoggerFactory _loggerFactory;
+    private readonly UriUtility _uriUtility;
+    private readonly IPublishedUrlProvider _publishedUrlProvider;
+
+    public ContentUrlFactory(
+        IPublishedRouter publishedRouter,
+        IUmbracoContextAccessor umbracoContextAccessor,
+        ILanguageService languageService,
+        ILocalizedTextService localizedTextService,
+        IContentService contentService,
+        IVariationContextAccessor variationContextAccessor,
+        ILoggerFactory loggerFactory,
+        UriUtility uriUtility,
+        IPublishedUrlProvider publishedUrlProvider)
+    {
+        _publishedRouter = publishedRouter;
+        _umbracoContextAccessor = umbracoContextAccessor;
+        _languageService = languageService;
+        _localizedTextService = localizedTextService;
+        _contentService = contentService;
+        _variationContextAccessor = variationContextAccessor;
+        _loggerFactory = loggerFactory;
+        _uriUtility = uriUtility;
+        _publishedUrlProvider = publishedUrlProvider;
+    }
+
+    public async Task<IEnumerable<ContentUrlInfo>> GetUrlsAsync(IContent content)
+    {
+        if (_umbracoContextAccessor.TryGetUmbracoContext(out IUmbracoContext? umbracoContext) == false)
+        {
+            _loggerFactory.CreateLogger<ContentUrlFactory>().LogWarning($"Unable to create an Umbraco context while attempting to get URLs for content: {content.Key}");
+            return Array.Empty<ContentUrlInfo>();
+        }
```

Comment:
I think it is okay to explode here, if the httpcontext is not there?
```suggestion
        IUmbracoContext umbracoContext = _umbracoContextAccessor.GetRequiredUmbracoContext();
```